### PR TITLE
Introduce prompflow-rag

### PR DESCRIPTION
**Pull request by jingyizhu99**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-index/README.md
Comment (wangchao1230) [inline src/promptflow-index/README.md:1]: add some description on what this pacakge does?

File: src/promptflow-index/setup.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+import os
+import re
+from io import open
+from typing import Any, List, Match, cast
+
+from setuptools import find_namespace_packages, setup
```
Comment (wangchao1230) [inline src/promptflow-index/setup.py:10]: add test for this pacakge?

File: src/promptflow-index/promptflow/index/resources/_index_data_source.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+    :param input_data: An input object describing the local location of index source files.
+    :type input_data: ~azure.ai.ml.Input
+    """
+
+    def __init__(self, *, input_data: str):  # todo Make sure type of input_data is correct
+        self.input_data = Input(type="uri_folder", path=input_data)
+        super().__init__(input_type=IndexInputType.LOCAL)
+
+    def _createComponent(self, index_config: IndexConfig, acs_config: Optional[AzureAISearchConfig] = None) -> Pipeline:
```
Comment (wangchao1230) [inline src/promptflow-index/promptflow/index/resources/_index_data_source.py:187]: this function name is not pythonic

File: src/promptflow-index/promptflow/index/__init__.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (wangchao1230) [inline src/promptflow-index/promptflow/index/__init__.py:2]: please address issues from lint CIs

File: src/promptflow-index/README.md
Comment (jingyizhu99) [reply src/promptflow-index/README.md:1]: yes, just did

File: src/promptflow-index/setup.py (hunk 1 above)
Comment (jingyizhu99) [reply src/promptflow-index/setup.py:10]: This is a known gap in our product, and we have work item to address this. To meet the current deadline for build, I will come back to add tests once done with the rest of the work for index sdk migration.

File: src/promptflow-index/promptflow/index/resources/_index_data_source.py (hunk 2 above)
Comment (jingyizhu99) [reply src/promptflow-index/promptflow/index/resources/_index_data_source.py]: thanks for catching that! Changed to `_create_component`

File: src/promptflow-index/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+azureml-rag[azure,faiss,cognitive_search,document_parsing]
+azure-ai-ml>=1.14.0
```
Comment (wangchao1230) [inline src/promptflow-index/requirements.txt:2]: not sure if we should separate azure depedency to extra?

Comment (ttthree) [reply src/promptflow-index/requirements.txt:2]: Isn't promptflow-index supposed to be "run locally" without AzureML dependency?
If it takes azure-ai-ml dependency anyway, maybe we should have the local index building in azure-ai-ml as well ...

Comment (ttthree) [reply src/promptflow-index/requirements.txt:2]: @danielsc thoughts?

**Review by ttthree — CHANGES_REQUESTED** [review]

File: src/promptflow-index/requirements.txt (hunk 4 above)
Comment (jingyizhu99) [reply src/promptflow-index/requirements.txt:2]: My bad, azure-ai-ml is not needed. Will remove it

Comment (eniac871) [reply src/promptflow-index/requirements.txt]: I don't think simply remove azure-ai-ml dependency works, azureml-rag[azure] still depends on azure-ai-ml. And from what I read, most of the index built method rely on Azure? _create_mlindex_from_existing_ai_search need to leverage mlclient to get connection, another path create_index_from_raw_embeddings totally rely on ACS to create an MLIndex.

File: src/promptflow-index/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+azureml-rag[azure,faiss,cognitive_search,document_parsing]
```
Comment (wangchao1230) [inline src/promptflow-index/requirements.txt:1]: Could you share a link to the code of azureml-rag?

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow-rag/setup.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+with open("README.md", encoding="utf-8") as f:
+    readme = f.read()
+
+with open("CHANGELOG.md", encoding="utf-8") as f:
+    changelog = f.read()
+
+setup(
+    name=PACKAGE_NAME,
+    version=version,
+    description="Prompt flow index build",
```
[the remark this answers is not in the document]
Comment (jingyizhu99) [reply src/promptflow-rag/setup.py:41]: This package does only contain index build for right now. We can update the description once we include index lookup tool in the future

File: src/promptflow-rag/promptflow/rag/_build_mlindex.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+        if not index_config.ai_search_connection_id:
+            import os
+
+            ai_search_args = {
+                **ai_search_args,
+                **{
+                    "endpoint": os.getenv("AZURE_AI_SEARCH_ENDPOINT")
+                    if "AZURE_AI_SEARCH_ENDPOINT" in os.environ
+                    else os.getenv("AZURE_COGNITIVE_SEARCH_TARGET"),
+                    "api_version": "2023-07-01-preview",
```
[the remark this answers is not in the document]
Comment (jingyizhu99) [reply src/promptflow-rag/promptflow/rag/_build_mlindex.py]: Make sense, changed to constant, but I notice there are other places in this repo that hard code api_version

File: src/promptflow-rag/setup.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        "Operating System :: OS Independent",
+    ],
+    python_requires="<4.0,>=3.8",
+    install_requires=parse_requirements('requirements.txt'),
+    packages=find_namespace_packages(include=[f"{PACKAGE_FOLDER_PATH}.*"]),
+    include_package_data=True,
+    project_urls={
+        "Bug Reports": "https://github.com/microsoft/promptflow/issues",
+        "Source": "https://github.com/microsoft/promptflow",
+    },
```
[the remark this answers is not in the document]
Comment (jingyizhu99) [reply src/promptflow-rag/setup.py:67]: Yes, this is an item in our backlog. We will prioritize it after build

File: .cspell.json [hunk 9]
```diff
@@ -198,10 +198,15 @@
     "wscript",
     "raisvc",
     "evals",
+    "mlindex",
```
Comment (wangchao1230) [inline .cspell.json:201]: conflict in this file

Comment (jingyizhu99) [reply .cspell.json:201]: thanks, conflict has been resolved

**Review by wangchao1230 — APPROVED** [review]

**Review by eniac871 — APPROVED** [review]

**Review by ttthree — APPROVED** [review]

**Pull request merged**